packages/api/AlvTime.Business/InvoiceRate/InvoiceRateService.cs

```diff
@@ -55,111 +56,145 @@ public async Task<decimal> GetEmployeeInvoiceRateForPeriod(DateTime fromDate, Da
-
-        var billableHours = GetMonthlyDataset(taskDictionary.ContainsKey(TaskType.BILLABLE) ? taskDictionary[TaskType.BILLABLE] : null, invoiceRatePeriodStart, invoiceRatePeriodEnd);
-        var nonBillableHours = GetMonthlyDataset(taskDictionary.ContainsKey(TaskType.NON_BILLABLE) ? taskDictionary[TaskType.NON_BILLABLE] : null, invoiceRatePeriodStart, invoiceRatePeriodEnd);
-        var vacationHours = GetMonthlyDataset(taskDictionary.ContainsKey(TaskType.VACATION) ? taskDictionary[TaskType.VACATION] : null, invoiceRatePeriodStart, invoiceRatePeriodEnd);
-
-        return new InvoiceStatisticsDto
-        {
-            BillableHours = billableHours,
-            NonBillableHours = nonBillableHours,
-            VacationHours = vacationHours,
-            InvoiceRate = GetInvoiceRateForArray(billableHours, vacationHours, invoiceRatePeriodStart, invoiceRatePeriodEnd),
-            NonBillableInvoiceRate = GetInvoiceRateForArray(nonBillableHours, vacationHours, invoiceRatePeriodStart, invoiceRatePeriodEnd),
-            Labels = GetMonthlyLabels(invoiceRatePeriodStart, billableHours.Count())
+        var invoicePeriodStart = GetInvoicePeriodStart(fromDate, invoicePeriod, extendPeriod);
+        var invoicePeriodEnd = GetInvoicePeriodEnd(toDate, invoicePeriod, extendPeriod);
+
+        var userTasks = await _timeRegistrationStorage.GetTimeEntriesWithCustomer(user.Id, invoicePeriodStart, invoicePeriodEnd);
+        var taskPeriodGrouping = GroupTasksByInvoicePeriod(userTasks, invoicePeriod);
+
+        var billableHours = new List<decimal>();
+        var nonBillableHours = new List<decimal>();
+        var vacationHours = new List<decimal>();
+        var invoiceRates = new List<decimal>();
+        var nonBillableInvoiceRate = new List<decimal>();
+        var starts = new List<DateTime>();
+        var ends = new List<DateTime>();
+
+        var previousPeriodEnd = invoicePeriodStart.AddSeconds(-1);
+        foreach (var grouping in taskPeriodGrouping)
+        {
+            var periodStart = (grouping.Key.periodStart < invoicePeriodStart) ? invoicePeriodStart : grouping.Key.periodStart;
+            var periodEnd = (grouping.Key.periodEnd > invoicePeriodEnd) ? invoicePeriodEnd : grouping.Key.periodEnd;
+
+            while (includeZeroPeriods && (periodStart - previousPeriodEnd).TotalSeconds != 1)
+            {
+                var start = previousPeriodEnd.AddSeconds(1);
+                var end = GetInvoicePeriodEnd(start, invoicePeriod, ExtendPeriod.End);
+
+                billableHours.Add(0);
+                nonBillableHours.Add(0);
+                vacationHours.Add(0);
+                invoiceRates.Add(0);
+                nonBillableInvoiceRate.Add(0);
+                starts.Add(start);
+                ends.Add(end);
+
+                previousPeriodEnd = end;
+            }
+
+            var bh = grouping.Where(timeEntry => GetTaskType(timeEntry) == TaskType.BILLABLE).Sum(timeEntry => timeEntry.Value);

```

Comment:
Consider writing `bh`, `nbh` and `vh` to something easier to read like `billableHours`, `nonBillableHours` and `vacationHours`. #selfDocumentingCode

Reply:
billableHours, nonBillableHours and vacationHours is already in use, so I need some alternative words. Do you have any suggestions. currentBillableHours, tempBillableHours, billableHours1?